core/src/main/java/com/sequenceiq/cloudbreak/controller/StackCreatorService.java

```diff
@@ -379,9 +382,43 @@ void fillInstanceMetadata(DetailedEnvironmentResponse environment, Stack stack)
                 instanceMetaData.setPrivateId(privateIdNumber++);
                 instanceMetaData.setInstanceStatus(InstanceStatus.REQUESTED);
             }
-            multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs, instanceGroup);
-            prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, instanceGroup);
         }
+        if (stack.getType() == StackType.DATALAKE) {
+            for (InstanceGroup instanceGroup : sortInstanceGroups(stack)) {
+                if (!(instanceGroup.getGroupName().equals(InstanceGroupName.GATEWAY) || instanceGroup.getGroupName().equals(InstanceGroupName.AUXILIARY))) {
+                    multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs, instanceGroup.getInstanceGroupNetwork(),
+                            instanceGroup.getNotDeletedAndNotZombieInstanceMetaDataSet());
+                }
+                prepareInstanceMetaDataSubnetAndAvailabilityZoneAndRackId(stackSubnetId, stackAz, instanceGroup);
+            }
+
+            Set<InstanceMetaData> mergedInstanceMetaData = new HashSet<>();
+            mergedInstanceMetaData.addAll(getInstanceMetadata(stack, InstanceGroupName.GATEWAY));
+            mergedInstanceMetaData.addAll(getInstanceMetadata(stack, InstanceGroupName.AUXILIARY));
+            multiAzCalculatorService.calculateByRoundRobin(subnetAzPairs,
+                    getInstanceGroupByName(stack, InstanceGroupName.GATEWAY).get().getInstanceGroupNetwork(), mergedInstanceMetaData);
+        } else {
```

Comment:
Could you please add logs for this logic?
There are multiple conditions without any line of logs.